optuna/integration/botorch.py

```diff
@@ -0,0 +1,468 @@
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+def _get_default_candidates_func(
+    n_objectives: int,
+) -> Callable[
+    [
+        "torch.Tensor",
+        "torch.Tensor",
+        Optional["torch.Tensor"],
+        "torch.Tensor",
+    ],
+    "torch.Tensor",
+]:
+    if n_objectives == 1:
+        return qei_candidates_func
+    elif n_objectives > 1:
+        return qehvi_candidates_func
+    else:
+        assert False, "Should not reach."
+
+
+# TODO(hvy): Allow utilizing GPUs via some parameter, not having to rewrite the callback
+# functions.
+@experimental("2.4.0")
+class BoTorchSampler(BaseMultiObjectiveSampler):
+    """A sampler that uses BoTorch, a Bayesian optimization library built on top of PyTorch.
+
+    This sampler allows using BoTorch's optimization algorithms from Optuna to suggest parameter
+    configurations. Parameters are transformed to continuous space and passed to BoTorch, and then
+    transformed back to Optuna's representations. Categorical parameters are one-hot encoded.
+
+    .. seealso::
+        See an `example <https://github.com/optuna/optuna/blob/master/examples/multi_objective/
+        botorch_simple.py>`_ how to use the sampler.
+
+    .. seealso::
+        See the `BoTorch <https://botorch.org/>`_ homepage for details and for how to implement
+        your own ``candidates_func``.
+
+    .. note::
+        An instance of this sampler *should be not used with different studies* when used with
+        constraints. Instead, a new instance should be created for each new study. The reason for
+        this is that the sampler is stateful keeping all the computed constraints.
+
+    Args:
+        candidates_func:
+            An optional function that suggests the next candidates. It must take the training
+            data, the objectives, the constraints, the search space bounds and return the next
+            candidates. The arguments are of type ``torch.Tensor``. The return value must be a
+            ``torch.Tensor``. However, if ``constraints_func`` is omitted, constraints will be
+            :obj:`None`.
+
+            If omitted, is determined automatically based on the number of objectives. If the
+            number of objectives is one, Quasi MC-based batch Expected Improvement (qEI) is used.
+            If the number of objectives is larger than one, Quasi MC-based batch Expected
+            Hypervolume Improvement (qEHVI) is used.
+
+            The function should assume *maximization* of the objective.
+
+            .. seealso::
+                See :func:`optuna.integration.botorch.qei_candidates_func` for an example.
+        constraints_func:
+            An optional function that computes the objective constraints. It must take a
+            :class:`~optuna.multi_objective.study.MultiObjectiveStudy`, a
+            :class:`~optuna.multi_objective.trial.FrozenMultiObjectiveTrial` and return the
```

Comment:
I aligned it with the callback to `study.optimize`. I'm not sure if it is needed. I do think it's easier to remember the APIs if these callbacks are aligned, but it's open for discussion.

Reply:
Hmm... Since `study.optimize` represents the entire optimization, I think it's natural for that callback to have `study`. On the other hand, since `BoTorchSampler.sample_relative` is a method that is only called in one trial, it seems unnatural for the callback to have `study`. I'd like to hear from other reviewers' opinions as well.

Reply:
Let me resolve this conversation since this is addressed.